src/plugin/connector/github_connector.py

```diff
@@ -2,52 +2,81 @@
 from typing import Dict, List
 from github import Github
 from spaceone.core.connector import BaseConnector
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from functools import lru_cache
 
 _LOGGER = logging.getLogger(__name__)
 
 class GithubConnector(BaseConnector):
     def __init__(self, github_access_token: str, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.client = Github(github_access_token)
+    
+    # 캐싱
+    @lru_cache(maxsize=128)
+    def list_repositories(self, actions: bool) -> List[Dict]:
+        return self._list_repositories(actions)
 
-    def list_repositories(self) -> List[Dict]:
+    def _list_repositories(self, actions: bool) -> List[Dict]:
         try:
             repos = self.client.get_user().get_repos()
             repo_list = []
-            for repo in repos:
-                repo_info = {
-                    'name': repo.name,
-                    'full_name': repo.full_name,
-                    'private': repo.private,
-                    'description': repo.description,
-                    'html_url': repo.html_url,
-                    'created_at': repo.created_at.isoformat(),
-                    'updated_at': repo.updated_at.isoformat(),
-                    'pushed_at': repo.pushed_at.isoformat(),
-                    'pull_requests': repo.get_pulls().totalCount,
-                    'branches': [branch.name for branch in repo.get_branches()],
-                    'workflows': self.get_workflows(repo),
-                    'actions': self.get_actions(repo)
-                }
-                repo_list.append(repo_info)
+
+            # ThreadPoolExecutor를 사용하여 병렬 처리
+            # 각 리포지토리에 대해 스레드 풀에서 비동기적으로 처리
+            with ThreadPoolExecutor(max_workers=10) as executor:
+                future_to_repo = {executor.submit(self.get_repo_info, repo, actions): repo for repo in repos}
+                # future: 스레드 풀에서 작업의 완료 상태나 결과를 확인할 수 있는 객체
+                for future in as_completed(future_to_repo): # 작업이 완료되는 순서대로 future 객체를 반환
+                    try:
+                        repo_info = future.result()
+                        if repo_info:
+                            repo_list.append(repo_info)
+                    except Exception as e:
+                        _LOGGER.error(f"Error fetching repository info: {e}")
+
             return repo_list
         except Exception as e:
             _LOGGER.error(f"Error fetching repositories from GitHub: {e}")
             return []
-
+    
+    
+    def get_repo_info(self, repo, fetch_actions: bool) -> Dict:
+        try:
+            actions_data = []
+            if fetch_actions:
+                actions_data = self.get_actions(repo)
+            
```

Comment:
actions에서는 branch에 관한 정보를 사용하지 않는 것으로 보이는 데요
조건문 내에서 가져오는게 어떨까요?


```suggestion

if fetch_actions:
           actions_data = self.get_actions(repo)
           branches = []
else:
           branches = branch.name for branch in repo.get_branches()],
```

Reply:
오... actions에는 brances 정보와 매핑이 되지 않으니 최적화에 더 기여할 것 같습니다. 적용해보겠습니다.

Reply:
`get_repo_info` 뿐만 아니라 `list_repositories`, `get_workflows` 에 `@lru_cache(maxsize=256)` 처럼 데코레이터 달아서 캐싱 처리할 수도 있을 것 같거든요. 

이전에 코드 보니까 `async` 키워드 이용해서 비동기 프로그래밍하신 것 같은데 안타깝게도 저희 Inventory 마이크로서비스가 비동기를 지원하지 않아서 Plugin 레벨에서는 비동기 코드를 작성할 수 없습니다.

다만 `@lru_cache` 이용해서 동기 프로그래밍에서 메서드 결과를 캐시하면 효과가 더 좋아질 것 같습니다. 여기서 `max_size` 파라미터를 조절해서 더 성능이 좋아지는 쪽으로 비교해봐도 될 것 같아요.

만약 네트워크 호출을 단 1번만 한다면 캐싱 처리가 필요없겠지만 2번 이상 호출한다면 굳이 API를 2번 찌를 필요는 없어보여요 

그리고 이 경우 데이터의 크기가 엄청 크거나 하지 않기때문에 드라마틱한 변화는 없을 수 있어도 한번 적용해보시고 성능 차이가 있다면 데코레이터를 이용해봐도 좋을 것 같습니다. 

`from functools import lru_cache` 하시고 메서드 위에 데코레이터 붙이시면 됩니다.

Reply:
아 참고로 이미 `list_repositories` 에서 캐싱 처리하셨지만 `_list_repositories`에서 멀티스레드 이용하고 있어서 이 [리뷰](https://github.com/ChainsoMen/plugin-github-inven-collector/pull/4/commits/2f8da272fa8ee707887e920495d486d3b305ccf5#r1802208589)처럼 멀티스레드를 빼고 캐싱만 해도 좋을 것 같다는 의미에서 말씀드렸습니다. 

사실 어떤 경우라도 더 성능 차이가 좋다면 멀티스레드를 그대로 이용하셔도 됩니다. 단, 현재 저희 cloudforet-io의 플러그인에서는 디버깅 및 성능 이슈로 멀티스레드를 이용하고 있진 않습니다.

Reply:
넵, 되도록이면 cloudforet-io 플러그인에서 멀티스레드 사용을 지양해보겠습니다.